### PR TITLE
Implementation of URLProtocol and refactoring URLSession

### PR DESCRIPTION
**Pull request by nethraravindran**
Currently URLSessionTask is tightly coupled with HTTP/HTTPS specific functions. This makes it difficult to introduce support for other native and custom protocols. Refactoring URLSession helps in adding support for other native protocols like ftp and data as well. Also, URLProtocol implementation provides support for custom protocols.
 
This document provides the proposal for restructuring URLSession
https://gist.github.com/pushkarnk/1bae3167d2e62b09ba0dd4b0dd285179

### CODE REVIEW

File: Foundation/NSURLProtocol.swift [hunk 1]
```diff
@@ -323,19 +433,54 @@ open class URLProtocol : NSObject {
         The only way that failure can occur is if the given class is not a
         subclass of NSURLProtocol.
     */
-    open class func registerClass(_ protocolClass: AnyClass) -> Bool { NSUnimplemented() }
-    
+    open class func registerClass(_ protocolClass: AnyClass) -> Bool {
+        if protocolClass is URLProtocol.Type {
+            guard !_registeredProtocolClasses.contains(where: { $0 === protocolClass }) else { return true }
+            _lock.lock()
```
Comment (ianpartridge) [inline Foundation/NSURLProtocol.swift]: I think this needs to go before the `guard` otherwise threads can still race?

File: Foundation/NSURLProtocol.swift [hunk 2]
```diff
@@ -323,19 +433,54 @@ open class URLProtocol : NSObject {
...
+            _registeredProtocolClasses.append(protocolClass)
+            _lock.unlock()
+            return true
+        }
+        return false
+    }
+
+    // TODO: Synchronize all the class funcs that use the registered protocol classes array
+    internal class func getProtocolClass(for urlRequest: URLRequest) -> AnyClass? {
```
Comment (ianpartridge) [inline Foundation/NSURLProtocol.swift]: Locking needed here too.

File: Foundation/NSURLProtocol.swift [hunk 3]
```diff
@@ -323,19 +433,54 @@ open class URLProtocol : NSObject {
...
+
     /*! 
         @method unregisterClass:
         @abstract This method unregisters a protocol. 
         @discussion After unregistration, a protocol class is no longer
         consulted in calls to NSURLProtocol class methods.
         @param protocolClass The class to unregister.
     */
-    open class func unregisterClass(_ protocolClass: AnyClass) { NSUnimplemented() }
+    open class func unregisterClass(_ protocolClass: AnyClass) {
```
Comment (ianpartridge) [inline Foundation/NSURLProtocol.swift]: Locking needed here.

File: Foundation/NSURLProtocol.swift [hunk 4]
```diff
@@ -151,7 +240,9 @@ public protocol URLProtocolClient : NSObjectProtocol {
     or more protocols or URL schemes.
 */
 open class URLProtocol : NSObject {
-    
+
+    private static var _registeredProtocolClasses = [AnyClass]()
+    private static var _lock = NSLock()
```
Comment (ianpartridge) [inline Foundation/NSURLProtocol.swift]: Perhaps call this `_classesLock`?

File: Foundation/NSURLProtocol.swift [hunk 5]
```diff
@@ -165,21 +256,34 @@ open class URLProtocol : NSObject {
...
-    public init(request: URLRequest, cachedResponse: CachedURLResponse?, client: URLProtocolClient?) { NSUnimplemented() }
-    
+    public required init(request: URLRequest, cachedResponse: CachedURLResponse?, client: URLProtocolClient?) {
+        self._request = request
+        self._cachedResponse = cachedResponse
+        self._client = client ?? _ProtocolClient()
+    }
+
+    private var _request : URLRequest
+    private var _cachedResponse : CachedURLResponse?
```
Comment (ianpartridge) [inline Foundation/NSURLProtocol.swift]: We don't seem to use this anywhere?

File: TestFoundation/main.swift [hunk 6]
```diff
@@ -80,7 +80,7 @@ XCTMain([
     testCase(TestNSURLResponse.allTests),
     testCase(TestNSHTTPURLResponse.allTests),
 //Disabling because of https://bugs.swift.org/browse/SR-4655 and https://bugs.swift.org/browse/SR-4647
```
Comment (pushkarnk) [inline TestFoundation/main.swift]: This comment needs to be removed if we are enabling the tests

Comment (ianpartridge) [inline TestFoundation/main.swift]: I don't think we should re-enable the tests under this PR.

File: Foundation/NSURLProtocol.swift [hunk 7]
```diff
@@ -142,6 +145,92 @@ public protocol URLProtocolClient : NSObjectProtocol {
...
+            guard let s = `protocol`.task?.session as? URLSession else { fatalError() }
+            s.delegateQueue.addOperation {
+                completion(data, `protocol`.task?.response, nil)
+                `protocol`.task?.state = .completed
+            }
+        default: return
+        }
+    }
+
+    func urlProtocol(_ protocol: URLProtocol, didFailWithError error: Error) {
```
Comment (pushkarnk) [inline Foundation/NSURLProtocol.swift]: nit: Will it be possible to factor out some code this function shares with `urlProtocol(_ protocol: URLProtocol, didFailWithError error: Error)`?

File: Foundation/NSURLProtocol.swift [hunk 8]
```diff
@@ -323,19 +435,62 @@ open class URLProtocol : NSObject {
...
         @method unregisterClass:
         @abstract This method unregisters a protocol. 
         @discussion After unregistration, a protocol class is no longer
         consulted in calls to NSURLProtocol class methods.
         @param protocolClass The class to unregister.
     */
-    open class func unregisterClass(_ protocolClass: AnyClass) { NSUnimplemented() }
+    open class func unregisterClass(_ protocolClass: AnyClass) {
+        if let idx = _registeredProtocolClasses.index(where: { $0 === protocolClass }) {
+            _classesLock.lock()
```
Comment (pushkarnk) [inline Foundation/NSURLProtocol.swift]: The lock acquisition needs to be raised above the `if`, given that `_registeredProtocolClasses` may be concurrently accessed

File: Foundation/NSURLSession/NSURLSession.swift [hunk 9]
```diff
@@ -219,6 +219,11 @@ open class URLSession : NSObject {
         let c = URLSession._Configuration(URLSessionConfiguration: configuration)
         self._configuration = c
         self.multiHandle = _MultiHandle(configuration: c, workQueue: workQueue)
+        // registering all the protocol classes with URLProtocol
+        let _ = URLProtocol.registerClass(_HTTPURLProtocol.self)
```
Comment (pushkarnk) [inline Foundation/NSURLSession/NSURLSession.swift]: I am wondering if we want to register the `native protocol classes` every time a new URLSession is created. This can be a one-time activity.

File: Foundation/NSURLSession/NSURLSession.swift [hunk 10]
```diff
@@ -234,6 +239,11 @@ open class URLSession : NSObject {
...
```
Comment (pushkarnk) [inline Foundation/NSURLSession/NSURLSession.swift]: I am wondering if we want to register the native protocol classes every time a new URLSession is created. This can be a one-time activity.

File: Foundation/NSURLSession/NSURLSessionTask.swift [hunk 11]
```diff
@@ -113,7 +77,11 @@ open class URLSessionTask : NSObject, NSCopying {
         _ = FileManager.default.createFile(atPath: fileName, contents: nil)
         self.tempFileURL = URL(fileURLWithPath: fileName)
         super.init()
-        self.easyHandle = _EasyHandle(delegate: self)
+        if let urlProtocolClass = URLProtocol.getProtocolClass(for: request) {
+            if urlProtocolClass is URLProtocol.Type {
+                self._protocol = (urlProtocolClass as! URLProtocol.Type).init(task: self, cachedResponse: nil, client: nil)
```
Comment (pushkarnk) [inline Foundation/NSURLSession/NSURLSessionTask.swift]: nit: Can we separate the downcast on a different line of code ? Though this is fine, I'm for better readability!

File: Foundation/NSURLSession/NSURLSessionTask.swift [hunk 12]
```diff
@@ -149,14 +117,20 @@ open class URLSessionTask : NSObject, NSCopying {
         set { taskAttributesIsolation.async(flags: .barrier) { self._response = newValue } }
     }
     fileprivate var _response: URLResponse? = nil
+    internal var setResponse : URLResponse? {
```
Comment (pushkarnk) [inline Foundation/NSURLSession/NSURLSessionTask.swift]: Can't we call this simply `response` ?

File: Foundation/NSURLSession/http/HTTPURLProtocol.swift [hunk 13]
```diff
@@ -0,0 +1,886 @@
...
+    override func startLoading() {
+        resume()
+    }
+
+    override func stopLoading() {
+        if task?.state == .suspended {
+            suspend()
+        } else {
+            self.internalState = .transferFailed
+            completeTask(withError: (self.task?.error)!)
```
Comment (pushkarnk) [inline Foundation/NSURLSession/http/HTTPURLProtocol.swift]: We need to avoid this unguarded unwrapping

File: Foundation/NSURLSession/http/HTTPURLProtocol.swift [hunk 14]
```diff
@@ -0,0 +1,886 @@
...
+}
+
+internal extension _HTTPURLProtocol {
+    /// Start a new transfer
+    func startNewTransfer(with request: URLRequest) {
+        task?.currentRequest = request
+        guard let url = request.url else { fatalError("No URL in request.") }
+
+        self.internalState = .transferReady(createTransferState(url: url, workQueue: (task?.workQueue)!))
+        if (task?.suspendCount)! < 1 {
```
Comment (pushkarnk) [inline Foundation/NSURLSession/http/HTTPURLProtocol.swift]: Will it be possible to avoid this unchecked unwrapping?

**Comment by pushkarnk** [comment]
@nethraravindran Since some new files have been created, I guess you'd need to update the `Foundation.xcodeproj/project.pbxproj` file and include those changes as well.

File: Foundation/NSURLProtocol.swift [hunk 15]
```diff
@@ -142,6 +145,92 @@ public protocol URLProtocolClient : NSObjectProtocol {
...
 
+internal class _ProtocolClient : NSObject, URLProtocolClient {
+
+    func urlProtocol(_ protocol: URLProtocol, didReceive response: URLResponse, cacheStoragePolicy policy: URLCache.StoragePolicy) {
+        `protocol`.task?.setResponse = response
+    }
+
+    func urlProtocolDidFinishLoading(_ protocol: URLProtocol) {
+        guard let session = `protocol`.task?.session as? URLSession else { fatalError() }
+        switch session.behaviour(for: `protocol`.task!) {
```
Comment (pushkarnk) [inline Foundation/NSURLProtocol.swift]: May be we should set a guard  for `protocol.task` and generate a `fatalError` if its is nil?  We can then access task in the remainder of the method without worries of nil :)

**Comment by pushkarnk** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
@swift-ci test

**Comment by ianpartridge** [comment]
@parkera We think this is ready for merge.  Can you give your view?  Thanks.

**Comment by parkera** [comment]
If you're happy with it, go ahead.

File: Foundation/NSURLProtocol.swift [hunk 16]
```diff
@@ -323,19 +439,65 @@ open class URLProtocol : NSObject {
         The only way that failure can occur is if the given class is not a
         subclass of NSURLProtocol.
     */
-    open class func registerClass(_ protocolClass: AnyClass) -> Bool { NSUnimplemented() }
-    
+    open class func registerClass(_ protocolClass: AnyClass) -> Bool {
+        if protocolClass is URLProtocol.Type {
+            _classesLock.lock()
+            guard !_registeredProtocolClasses.contains(where: { $0 === protocolClass }) else {
+                _classesLock.unlock()
```
Comment (weissi) [inline Foundation/NSURLProtocol.swift]: I'd use

    defer {
        _classesLock.unlock()
    }

here and in the following cases, otherwise it looks quite easy that a further refactoring forgets to unlock...

**Comment by ianpartridge** [comment]
Thanks @parkera we're still looking at the `URLSession` test failures that showed up in the CI so will wait until we understand those a bit better.

File: Foundation/NSURLProtocol.swift (hunk 16 above)
Comment (pushkarnk) [inline Foundation/NSURLProtocol.swift]: I rememberwhat @parkera had to say about this :)
https://github.com/apple/swift-corelibs-foundation/pull/902#issuecomment-283996792

Comment (weissi) [inline Foundation/NSURLProtocol.swift]: very surprising but fair enough, thanks for pointing that out 👍

File: Foundation/NSURLSession/http/HTTPURLProtocol.swift [hunk 17]
```diff
@@ -0,0 +1,892 @@
...
+
+        easyHandle.set(customHeaders: customHeaders)
+
+        //TODO: The CURLOPT_PIPEDWAIT option is unavailable on Ubuntu 14.04 (libcurl 7.36)
+        //TODO: Introduce something like an #if, if we want to set them here
+
+        //set the request timeout
+        //TODO: the timeout value needs to be reset on every data transfer
+
+        var timeoutInterval = Int(httpSession.configuration.timeoutIntervalForRequest) * 1000
```
Comment (pushkarnk) [inline Foundation/NSURLSession/http/HTTPURLProtocol.swift]: This timeout logic needs to be moved to URLSessionTask since it is not something specific to HTTP. We can do this in a separate PR though.

File: Foundation/NSURLSession/NSURLSessionTask.swift [hunk 18]
```diff
@@ -26,55 +26,19 @@ import Dispatch
...
 open class URLSessionTask : NSObject, NSCopying {
     /// How many times the task has been suspended, 0 indicating a running task.
-    fileprivate var suspendCount = 1
-    fileprivate var easyHandle: _EasyHandle!
-    fileprivate var totalDownloaded = 0
-    fileprivate var session: URLSessionProtocol! //change to nil when task completes
-    fileprivate let body: _Body
-    fileprivate let tempFileURL: URL
+    internal var suspendCount = 1
+    internal var totalDownloaded = 0
```
Comment (pushkarnk) [inline Foundation/NSURLSession/NSURLSessionTask.swift]: This is specific to the `http download tasks`. We'd need to consider moving it to the HTTP implementation.

**Comment by pushkarnk** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
cc @nethraravindran build failures due to some compilation errors and warning

**Comment by pushkarnk** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
Another TODO: There's a lot of places where we're doing optional chaining. I understand that we will not have nil values at these points under normal circumstances. But for those abnormal nils that may come up once in a while, we have to define clear code paths (may be a guard with`fatalError()`). 

Since the tests are passing and the code changes are big, I am fine with merging this now. But it does warrant some future improvement work. Thanks!

File: Foundation/NSURLSession/NSURLSessionTask.swift (hunk 18 above)
Comment (nethraravindran) [inline Foundation/NSURLSession/NSURLSessionTask.swift]: Raised PR: https://github.com/apple/swift-corelibs-foundation/pull/1029

File: Foundation/NSURLSession/NSURLSession.swift (hunk 9 above)
Comment (nethraravindran) [inline Foundation/NSURLSession/NSURLSession.swift]: PR: https://github.com/apple/swift-corelibs-foundation/pull/1030

File: Foundation/NSURLSession/NSURLSession.swift (hunk 10 above)
Comment (nethraravindran) [inline Foundation/NSURLSession/NSURLSession.swift]: PR: https://github.com/apple/swift-corelibs-foundation/pull/1030
